### ISSUE TITLE
Passing test arguments for a project (e.g. numpy)

### ISSUE THREAD

**Issue by rgommers** [scientific-python/devpy#13]
The default `test` command allows for passing flags on to `pytest with `./dev.py test -- -custom-pytest-flag-here`, but it seems like there are no arguments that are implemented in the project. I don't see for example how to run the full numpy test suite, something that can be done with other interfaces like so:
- `python -c "import numpy as np; np.test('full')`
- `python runtests.py -m full`

In SciPy it's similar, `python dev.py test -m full`.

**Comment by stefanv** [comment]
I'll try to replicate on NumPy; I'm having issues building it (I filed it on the project board).

When I am in the skimage repo, I can do, e.g.:

```
./dev.py test -- --collect-only
```

So I was expecing `./dev.py test -- -m slow` to work.

What are you expecting to see, vs what is happening?

**Comment by rgommers** [comment]
Argh, this was 10+ years of muscle memory. NumPy uses `fast` (default) and `full`, not `slow` and `"not slow"`. `./dev.py test -- -m "not slow"` does work.

There's a design issue here though in `devpy` I think. The default is not the right one (it should be `not slow`), and that's not so easy to work around. If we add `-m "not slow`" to the default test invocation, then I expect a conflict when one _does_ want to run in "full" mode. And when you want to add more customization (e.g. run in parallel by default if `pytest-xdist` is installed) then that becomes more complex.

**Comment by stefanv** [comment]
The way it works right now, `./dev.py test` is shorthand for "run pytest with the PYTHONPATH set up correctly", similar to how shell and ipython work.

I can see at least two options: either numpy implements its own `test` command, or we add a `-m` flag to the `devpy` version, like what SciPy does. Then you can do:

```
./dev.py test  # runs with `-m "not slow"`
./dev.py -m ""  # runs without `-m` flag
./dev.py -- -m foo  # broken? runs with both -m "not slow" and -m foo
```

I don't know how standard `-m "not slow"` is. Standard enough to include in `devpy`? I don't think scikit-image uses it, e.g.

**Comment by rgommers** [comment]
It's probably not standard enough to include by default - that'd be odd.

I think for NumPy we're anyway going to want the full `dev.py` version like SciPy, there's too many custom things. I'll exercise `devpy` in the `meson` branch as much as possible, but the difference in experience with SciPy is large - and we don't want to keep adding bells and whistles in this repo until the two become almost equivalent.

**Comment by stefanv** [comment]
Ultimately, I *hope* there won't be a SciPy dev.py and a "not full devpy"? The idea is to have the tool be generalizable and easy to adopt by many different projects. I've just gone through the process of adopting it for skimage, and only had to implement a small handful of commands to replace our old makefile.

I guess the whole point is to have a central tool we can maintain, instead of each project vendoring its own `dev.py` script, from which improvements are hard to extract for broad consumption.

https://github.com/scikit-image/scikit-image/pull/6600/files#diff-da8e907803baf703c3c7d49530dddddd3c20dda146d8ed8a8a565a31f0aea10b

If there are any crucial features missing that should be adopted from the scipy dev.py, I'm happy to implement them.

**Comment by rgommers** [comment]
> If there are any crucial features missing that should be adopted from the scipy dev.py, I'm happy to implement them.

A couple of things that come to mind:

- It's unclear how to add project-specific functionality like extra flags for commands, as well as docs that are project-specific
- There are no dependencies between commands; `./dev.py test` does not run `build`, making usage more verbose and easier to get wrong
- The commands executed aren't echoed to the terminal (good for teaching users what is happening)
- The CLI is much less nice (groups, colors)

> I guess the whole point is to have a central tool we can maintain, instead of each project vendoring its own `dev.py` script, from which improvements are hard to extract for broad consumption.

It is nice to have a base version to work from, that is an easy drop-in and at least equal to or better than a makefile or the old `runtests.py`. But it's at that level now, and I think to get from there to SciPy's dev.py level there's a big gap. So I think there's a place for both - unless this one gains at least most of the missing features. The one downside I can think of for the SciPy version is the extra dependencies (but some are hard to avoid if you want the bells and whistles). 

I am still a little unsure why you'd start from scratch rather than refactor SciPy's version if you do aim for feature parity.

**Comment by stefanv** [comment]
Thanks for these helpful comments, Ralf!

> A couple of things that come to mind:
> 
>     * It's unclear how to add project-specific functionality like extra flags for commands, as well as docs that are project-specific

Agreed. The commands are simply functions, so if we get to the point where the devpy command API is stable enough, projects can start using the devpy commands inside their own commands, adding custom flags as necessary.

>     * There are no dependencies between commands; `./dev.py test` does not run `build`, making usage more verbose and easier to get wrong

Since these are just functions, in principle I see no problem calling one command from another.

>     * The commands executed aren't echoed to the terminal (good for teaching users what is happening)

They are; the `run` command used internally prints them with `$ cmd`. For example, the first output line of `./dev.py build` shows me:

```
$ meson setup /home/stefan/src/numpy/build --prefix=/usr --reconfigure
```

I can replace the $ with an emoji if that would make it clearer?

>     * The CLI is much less nice (groups, colors)

True, the goal was to implement something basic with a minimal set of dependencies, see what we truly miss, and add that back in. I think groups, especially, can be very helpful in organizing commands. Unfortunately, `click` doesn't support those out-of-the-box, as far as I can tell.

> I am still a little unsure why you'd start from scratch rather than refactor SciPy's version if you do aim for feature parity.

A good question, and one that warrants explanation! My goal was certainly not to rebuild it from scratch, and the first version indeed simply took the `dev.py` commits out of SciPy. However, I ran into some difficulties, mainly related to generalization and customizability. I wanted to make it *as easy as possible* for project maintainers to add new commands. The contexts used by SciPy's dev.py made the command definition and execution quite a bit harder to follow than plain functions. Also, as you mentioned, that rich, colorful experience, comes at the cost of additional dependencies.

So, I tried to build a tool that favored simple command authoring, using a minimal set of dependencies, that could get us most of the way there. I am not opposed to extending the set of dependencies, but I wanted to make very sure that we needed them before doing so.

Along those same lines, for the default commands provided, `devpy` has to ask some additional questions: will this be useful across projects, and what would a generic, yet useful, interface look like? With the obvious caveat that such an interface may not readily exist, in which case the project would have to build a custom command (hence the need to make that process entirely straightforward).

Obviously, having SciPy dev.py was extremely valuable in seeing how this all should come together! I should add a section to the README making that history clear.

**Comment by rgommers** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> I can replace the $ with an emoji if that would make it clearer?

Ah yes, I overlooked it. An emoji would be nice I think. In case we add back colors, using a different one (like light-gray/blue in SciPy) helps too.

**Comment by rgommers** [comment]
> that rich, colorful experience, comes at the cost of additional dependencies.

Maybe make that opt-in? It does add a lot of additional value I think. Making it pleasant to use can be a difference between adoption or having it only used by a smaller part of a team. I certainly avoided `runtests.py`, because it added a layer of complexity without adding enough value. While I quite happily use `dev.py`.

**Comment by stefanv** [comment]
A few issues that came out of this discussion have been closed. I didn't see any specific requests for the `test` command, but happy to add those as they come up.

**Comment by rgommers** [comment]
Thanks! If I may add one more that was fiddly today: `./dev.py doc` would be nice to have. Messing with PYTHONPATH and remembering to re-run `./dev.py build` when trying to make Sphinx happy is slightly non-ideal.

**Comment by stefanv** [comment]
Sure! The one in scikit-image (1) sets the PYTHONPATH (2) changes into doc and does "make html". Do you think that is generally useful enough?

See https://github.com/scikit-image/scikit-image/blob/main/.devpy/cmds.py#L16

**Comment by rgommers** [comment]
Yes, that should cover the majority of needs I'd think
